stream_alert/rule_processor/sink.py

```diff
@@ -92,3 +95,88 @@ def sink(self, alerts):
+        """Group alerts into batches of 25, the maximum allowed by Dynamo batch_write_item.
+
+        Yields:
+            (dict) The constructed request for batch_write_item, containing <= 25 alerts.
+                Maps table name to a list of requests.
+        """
+        for i in xrange(0, len(alerts), batch_size):
+            batch = alerts[i:i+batch_size]
+            yield {
+                self.table: [
+                    {
+                        'PutRequest': {
+                            'Item': {
+                                'RuleName': {'S': alert['rule_name']},
+                                'Timestamp': {
+                                    # ISO 8601 datetime format, and is unique for each alert
+                                    'S': datetime.now().strftime('%Y-%m-%dT%H:%M:%S.%fZ')
+                                },
+                                'Cluster': {'S': os.environ['CLUSTER']},
+                                'RuleDescription': {'S': alert['rule_description']},
+                                'Outputs': {'SS': alert['outputs']},
+                                # Compact JSON encoding (no extra spaces)
+                                'Record': {'S': json.dumps(alert['record'], separators=(',', ':'))}
+                            }
+                        }
+                    }
+                    for alert in batch
+                ]
+            }
+
+    @backoff.on_exception(backoff.expo, ClientError, max_tries=5, jitter=backoff.full_jitter,
+                          on_backoff=backoff_handlers.backoff_handler,
+                          on_success=backoff_handlers.success_handler,
+                          on_giveup=backoff_handlers.giveup_handler)
+    def _batch_write(self, request_items, max_attempts=5):
+        """Write a batch of alerts to Dynamo, retrying with exponential backoff for failed items.
+
+        Args:
+            request_items (dict): The generated RequestItems dict from _alert_batches().
+            max_attempts (int): Maximum number of times to retry UnprocessedItems.
+
+        Returns:
+            (bool) True if the batch write was eventually successful, False otherwise.
+        """
+        for attempt in xrange(1, max_attempts + 1):
+            response = self.client_dynamo.batch_write_item(RequestItems=request_items)
+
+            # If Dynamo experiences an internal error, unprocessed items are listed in the response.
+            # AWS recommends retrying unprocessed items in a loop with exponential backoff.
+            request_items = response['UnprocessedItems']
+            if request_items:
+                LOGGER.warn(
+                    'Batch write failed: %d alerts were not written (attempt %d/%d)',
+                    len(request_items[self.table]), attempt, max_attempts)
+                # Simple exponential backoff: Sleep 0.5, 1, 2, 4 and 8 seconds.
+                time.sleep(0.25 * 2 ** attempt)
```

Comment:
Austin explained offline - no need to comment here

Reply:
There are 2 different kinds of failures here:
1. If `Dynamo` is unavailable, a boto `ClientError` is retried in the backoff.
2. If some (but not all) of the records are not processed for some reason, these are returned in `UnprocessedItems`. I retried this latter one myself because the items being passed to `batch_write` change between successive retries.

After looking at this again, a much simpler solution would be to use `backoff` for both error cases and just store the `unprocessed_items` in the class.

Reply:
I'll also add some comments, since this is definitely confusing